testing-common/src/main/java/io/opentelemetry/instrumentation/testing/junit/AutoCleanupExtension.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.testing.junit;
+
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Queue;
+import org.junit.jupiter.api.extension.AfterEachCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
+
+/**
+ * A small utility extension that allows deferring executing cleanup code in {@code @Test} methods
+ * until the test has finished. All cleanup callbacks added during the test will always be executed
+ * after it finishes, no matter the outcome. Inspired by Spock's {@code cleanup:} block.
+ */
+public final class AutoCleanupExtension implements AfterEachCallback {
```

Comment:
I don't think we'd want to publish a helper that's not really related to instrumentation to end users, but we could split out, or maybe inline, later. What do you think?

Reply:
Hmm, good point - I'll move it to an `internal` package for now, we could split out a "shared internal testing code" module later if there's any need.